lib/middlewares/recurring.js

```diff
@@ -0,0 +1,120 @@
+"use strict"
+
+var Mongo = require('./mongo')
+var co = require('co')
+var shortid = require('shortid')
+
+class Recurring extends Mongo {
+  constructor(options) {
+    super(options)
+
+    this.defineMandatoryDependency('in_queue', 'inQueue', this.service)
+    this.defineMandatoryDependency('lock_manager', 'lockManager', this.service)
+
+    const FIVE_SECONDS = 5 * 1000
+    const ONE_MINUTE = 60 * 1000
+    this.pollingIntervalMillis = this.params.pollingIntervalMillis || FIVE_SECONDS
+    this.schedulingThresholdMillis = this.params.schedulingThresholdMillis || ONE_MINUTE
+    this.lockResourceId = `${this.service.name}_recurring_scheduling_lock`
+    this.collectionName = `${this.service.name}_recurring`
+  }
+
+  *initialize() {
+    yield *super.initialize()
+
+    this.collection = this.db.collection(this.collectionName)
+    yield this.collection.createIndex(
+      {recurringMessageId: 1},
+      {unique: true, name: "recurringMessageId"}
+    )
+    yield this.collection.createIndex(
+      {nextAt: 1},
+      {unique: false, name: "nextAt"}
+    )
+  }
+
+  *register(msg) {
+    var now = Date.now()
+    var record = {
+      _id: msg.messageType,     // prevent registering multiple messages of the same type
+      recurringMessageId: shortid.generate(),
+      message: msg,
+      registeredAt: now,
+      nextAt: now,
+      prevAt: null,
+    }
+    var existingMsg = yield this.collection.findOneAsync({_id: msg.messageType})
+    if (existingMsg) {
```

Comment:
Can't use it as it updates the whole document, while I need to update only the message part if it already exists (can't touch nextAt and prevAt)